File: common/src/main/java/com/ssblur/scriptor/blockentity/renderers/RuneBlockEntityRenderer.java
package com.ssblur.scriptor.blockentity.renderers;

import com.mojang.blaze3d.vertex.DefaultVertexFormat;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.blaze3d.vertex.VertexFormat;
import com.ssblur.scriptor.ScriptorMod;
import com.ssblur.scriptor.blockentity.RuneBlockEntity;
import com.ssblur.scriptor.color.CustomColors;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.RenderStateShard;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.resources.ResourceLocation;

public class RuneBlockEntityRenderer implements BlockEntityRenderer<RuneBlockEntity> {
  static ResourceLocation magicCircle = ScriptorMod.location("textures/entity/magic_circle.png");
  static ResourceLocation emptyCircle = ScriptorMod.location("textures/entity/empty_circle.png");
  static RenderType magicLayer =
    RenderType
      .create(
        ScriptorMod.MOD_ID + ":rune",
        DefaultVertexFormat.POSITION_COLOR_TEX_LIGHTMAP,
        VertexFormat.Mode.QUADS,
        64,
        false,
        false,
        RenderType.CompositeState.builder().setTextureState(new RenderStateShard.TextureStateShard(magicCircle, false, false))
          .setCullState(new RenderStateShard.CullStateShard(false))
          .setLightmapState(new RenderStateShard.LightmapStateShard(true))
          .setShaderState(RenderStateShard.POSITION_COLOR_TEX_LIGHTMAP_SHADER)
          .createCompositeState(true)
      );
  static RenderType emptyLayer =
    RenderType
      .create(
        ScriptorMod.MOD_ID + ":circle",
        DefaultVertexFormat.POSITION_COLOR_TEX_LIGHTMAP,
        VertexFormat.Mode.QUADS,
        64,
        false,
        false,
        RenderType.CompositeState.builder().setTextureState(new RenderStateShard.TextureStateShard(emptyCircle, false, false))
          .setCullState(new RenderStateShard.CullStateShard(false))
          .setLightmapState(new RenderStateShard.LightmapStateShard(true))
          .setShaderState(RenderStateShard.POSITION_COLOR_TEX_LIGHTMAP_SHADER)
          .createCompositeState(true)
      );

  public RuneBlockEntityRenderer(BlockEntityRendererProvider.Context context) {

  }
  @Override
  public void render(RuneBlockEntity rune, float tickDelta, PoseStack matrix, MultiBufferSource buffers, int light, int j) {
    if(rune.getLevel() == null) return;
    matrix.pushPose();

    int c = CustomColors.getColor(rune.color, rune.getLevel().getGameTime() + tickDelta);
    int r, g, b;
    var mc = Minecraft.getInstance();
    assert mc.level != null;
    r = (c & 0xff0000) >> 16;
    g = (c & 0x00ff00) >> 8;
    b = c & 0x0000ff;
    float yo = 0.0625f;
    matrix.translate(0, yo, 0);
    var pose = matrix.last().pose();
    var buffer = buffers.getBuffer(magicLayer);

    int cycle = 240;
    double rot = rune.getLevel().getGameTime() % cycle;
    rot = ((rot + tickDelta) / cycle) * 2 * Math.PI;
    double R = 0.5;
    float x = (float) (R * Math.cos(rot));
    float y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(0, 1).setLight(0xF000F0);
    rot += (Math.PI / 2);
    x = (float) (R * Math.cos(rot));
    y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(1, 1).setLight(0xF000F0);
    rot += (Math.PI / 2);
    x = (float) (R * Math.cos(rot));
    y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(1, 0).setLight(0xF000F0);
    rot += (Math.PI / 2);
    x = (float) (R * Math.cos(rot));
    y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(0, 0).setLight(0xF000F0);

    buffer = buffers.getBuffer(emptyLayer);
    R = 0.2;
    rot *= -1;
    x = (float) (R * Math.cos(rot));
    y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(0, 1).setLight(0xF000F0);
    rot += (Math.PI / 2);
    x = (float) (R * Math.cos(rot));
    y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(1, 1).setLight(0xF000F0);
    rot += (Math.PI / 2);
    x = (float) (R * Math.cos(rot));
    y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(1, 0).setLight(0xF000F0);
    rot += (Math.PI / 2);
    x = (float) (R * Math.cos(rot));
    y = (float) (R * Math.sin(rot));
    buffer.addVertex(pose, x + 0.5f, 0, y + 0.5f).setColor(r, g, b, 255).setUv(0, 0).setLight(0xF000F0);

    matrix.popPose();
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/blockentity/renderers/RuneBlockEntityRenderer.kt
package com.ssblur.scriptor.blockentity.renderers

import com.mojang.blaze3d.vertex.DefaultVertexFormat
import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.blaze3d.vertex.VertexFormat
import com.ssblur.scriptor.ScriptorMod
import com.ssblur.scriptor.blockentity.RuneBlockEntity
import com.ssblur.scriptor.color.CustomColors.getColor
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderStateShard
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.resources.ResourceLocation
import kotlin.math.cos
import kotlin.math.sin

class RuneBlockEntityRenderer(@Suppress("unused_parameter") context: BlockEntityRendererProvider.Context?):
  BlockEntityRenderer<RuneBlockEntity> {
  override fun render(
    rune: RuneBlockEntity,
    tickDelta: Float,
    matrix: PoseStack,
    buffers: MultiBufferSource,
    light: Int,
    j: Int
  ) {
    if (rune.level == null) return
    @Suppress("SENSELESS_COMPARISON")
    if (matrix == null) return
    matrix.pushPose()

    val c = getColor(rune.runeColor, rune.level!!.gameTime + tickDelta)
    val mc = Minecraft.getInstance()
    checkNotNull(mc.level)
    val r = c and 0xff0000 shr 16
    val g = c and 0x00ff00 shr 8
    val b = c and 0x0000ff
    val yo = 0.0625f
    matrix.translate(0f, yo, 0f)
    val pose = matrix.last().pose()
    var buffer = buffers.getBuffer(magicLayer)

    val cycle = 240
    var rot = (rune.level!!.gameTime % cycle).toDouble()
    rot = ((rot + tickDelta) / cycle) * 2 * Math.PI
    var R = 0.5
    var x = (R * cos(rot)).toFloat()
    var y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(0f, 1f).setLight(0xF000F0)
    rot += (Math.PI / 2)
    x = (R * cos(rot)).toFloat()
    y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(1f, 1f).setLight(0xF000F0)
    rot += (Math.PI / 2)
    x = (R * cos(rot)).toFloat()
    y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(1f, 0f).setLight(0xF000F0)
    rot += (Math.PI / 2)
    x = (R * cos(rot)).toFloat()
    y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(0f, 0f).setLight(0xF000F0)

    buffer = buffers.getBuffer(emptyLayer)
    R = 0.2
    rot *= -1.0
    x = (R * cos(rot)).toFloat()
    y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(0f, 1f).setLight(0xF000F0)
    rot += (Math.PI / 2)
    x = (R * cos(rot)).toFloat()
    y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(1f, 1f).setLight(0xF000F0)
    rot += (Math.PI / 2)
    x = (R * cos(rot)).toFloat()
    y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(1f, 0f).setLight(0xF000F0)
    rot += (Math.PI / 2)
    x = (R * cos(rot)).toFloat()
    y = (R * sin(rot)).toFloat()
    buffer.addVertex(pose, x + 0.5f, 0f, y + 0.5f).setColor(r, g, b, 255).setUv(0f, 0f).setLight(0xF000F0)

    matrix.popPose()
  }

  companion object {
    var magicCircle: ResourceLocation = ScriptorMod.location("textures/entity/magic_circle.png")
    var emptyCircle: ResourceLocation = ScriptorMod.location("textures/entity/empty_circle.png")
    var magicLayer: RenderType = RenderType
      .create(
        ScriptorMod.MOD_ID + ":rune",
        DefaultVertexFormat.POSITION_COLOR_TEX_LIGHTMAP,
        VertexFormat.Mode.QUADS,
        64,
        false,
        false,
        RenderType.CompositeState.builder()
          .setTextureState(RenderStateShard.TextureStateShard(magicCircle, false, false))
          .setCullState(RenderStateShard.CullStateShard(false))
          .setLightmapState(RenderStateShard.LightmapStateShard(true))
          .setShaderState(RenderStateShard.POSITION_COLOR_TEX_LIGHTMAP_SHADER)
          .createCompositeState(true)
      )
    var emptyLayer: RenderType = RenderType
      .create(
        ScriptorMod.MOD_ID + ":circle",
        DefaultVertexFormat.POSITION_COLOR_TEX_LIGHTMAP,
        VertexFormat.Mode.QUADS,
        64,
        false,
        false,
        RenderType.CompositeState.builder()
          .setTextureState(RenderStateShard.TextureStateShard(emptyCircle, false, false))
          .setCullState(RenderStateShard.CullStateShard(false))
          .setLightmapState(RenderStateShard.LightmapStateShard(true))
          .setShaderState(RenderStateShard.POSITION_COLOR_TEX_LIGHTMAP_SHADER)
          .createCompositeState(true)
      )
  }
}
